abstract/imaging_device.py

```diff
@@ -1,14 +1,121 @@
 # extended from https://gist.github.com/HazenBabcock/8d1042f564dcaee60c3a3972f2d56999
 
 
+class HWPropertyValueError(Exception):
+    pass
+
+
+class HWProperty(object):
+    """Defines a hardware property.
+    """
+    def __init__(self,
+                 name = None,
+                 value = None,
```

Comment:
I'm thinking about this some more and wondering whether `value` does not belong here. I can definitely see using HWProperty as a way to describe the available properties and their limits, but it makes less sense to me as a way of getting/setting values on the device. I've generally found that I only request the property metadata when setting up user interfaces, and during actual acquisition I always just work with {property:value} dictionaries. There's also this issue that `get_properties` is redundant to `HWProperty.value`, except that the latter method cannot easily be made plural (I'd prefer there to be only one way to get/set property values).  Thoughts?

Reply:
In my mental model, this is addressing that a call to `set_properties` could change the allowed ranges, allowed values or even the value of one or more properties. I think there are two ways to handle how this flows back to the UI.

1. The device and the interface share the same `HWProperty` object so any updates happen "automatically". This I think is what you are saying regarding `HWProperty.value` and `get_properties` being redundant.
2. Each time properties change the device creates a updated list of `HWProperty` objects. In this case the `HWProperty.value` reflects the current state of the device, but changing it's value has no effect unless `set_properties` is called. The UI is responsible for updating it's model of the device state based on the new list of `HWProperty` objects, any old `HWProperty` objects that have the same names as those in the new list will be stale.

I'm in favor of (2) even though there will be a lot more churn of `HWProperty` objects because it breaks the implicit connection created by shared objects representing device state which I think will make debugging much harder.

Reply:
Currently `set_properties` returns a dictionary of changed properties like `{'property': value}` and a list of the names of properties that might have their limits or other state changed, so I think that covers all of the information that needs to flow back to the UI, without the need for having `HWProperty.value`. One benefit of this approach is that it's light--we defer the effort of constructing those new HWProperty instances by just returning the names of properties that could have their limits changed.

Reply:
Okay, however we still need to have `HWProperty.value`, or else `get_properties` would need to change. The current proposal for `get_properties` is that it returns a dictionary of `HWProperty`.

Reply:
I would suggest `get_properties` returns `{'prop_name': value, ...}`, as originally specified.

Reply:
Okay, so forget the `HWProperty` idea entirely? Not sure how you get a `HWProperty` object if it isn't return by either `get_properties` or `set_properties`.

Reply:
Current description says `get_property_info` returns `{'prop_name': HWProperty}`

Reply:
As described it also says it would return information about all the properties supported by the device.

Reply:
Ah, so maybe it should be `get_property_info(list_of_prop_names)` ?

Reply:
And an empty list? `None`? Would return a list of all the properties. Works for me.